### PR TITLE
add win_delay_load_hook build process for electron on windows

### PR DESCRIPTION
**Pull request by Ryanmtate**
This PR resolves issue #17 by adding a build step for compiling a modified version of `win_delay_load_hook.cc`. The built artifact is linked during electron runtime.

This has been tested on Windows 10.

Screen Shot 2020-08-03 at 10 45 53 PM

In the future we may want to write this as a FFI to the WINAPI. In the short-term, hopefully this will help a few individuals using windows + electron + node-bindgen.

### CODE REVIEW

File: nj-build/Cargo.toml [hunk 1]
```diff
@@ -8,4 +8,8 @@ repository = "https://github.com/infinyon/node-bindgen"
 license = "Apache-2.0"
 
 [target.'cfg(windows)'.dependencies]
-reqwest = { version = "0.10", features = ["native-tls", "blocking"] }
\ No newline at end of file
+reqwest = { version = "0.10", features = ["native-tls", "blocking"] }
+cc = "1.0.58"
+
+[dependencies]
```
Comment (Ryanmtate) [inline nj-build/Cargo.toml:14]: remove

File: nj-build/src/win_delay_load_hook.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        let mut file = File::create(&tmp_file)?;
+        file.write_all(WIN_DELAY_LOAD_HOOK.as_bytes())?;
+    }
+
+    cc::Build::new()
+        .file(&tmp_file)
+        .out_dir(&dir)
+        .compile(file_name);
+    println!("cargo:rerun-if-changed={}", tmp_file.display());
```
Comment (Ryanmtate) [inline nj-build/src/win_delay_load_hook.rs:63]: check if necessary, remove if not needed.

**Comment by sehz** [comment]
Great work.   I guess only downside is dependency to cc.   Can you try to see if you can port that window_delay_hook.c as Rust FFI code?  It allows also to customize code later easily as situation changes.

**Comment by Ryanmtate** [comment]
> Great work. I guess only downside is dependency to cc. Can you try to see if you can port that window_delay_hook.c as Rust FFI code? It allows also to customize code later easily as situation changes.

I'll look into using rust-bindgen to generate the `windows.h` and other header files. We can either hold off on this PR until that is finished, or merge this in as a partial fix. While it does introduce an external dependency, at least the license on cc-rs is MIT.

**Comment by sehz** [comment]
Ok.  Let's use this solution as temporary fix and create another issue to migrate off cc.   There is also https://github.com/retep998/winapi-rs, that you might want to check it out to see if it can be used for windows binding.

File: nj-build/src/lib.rs [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
 
-    let node_full_version = String::from_utf8(Command::new("node").arg("-v").output().unwrap().stdout).unwrap();
-        
-    let temp_dir =   PathBuf::from("C:\\Windows\\Temp\\");
-    let mut temp_lib =   temp_dir.clone();
-    
+    let node_full_version =
+        String::from_utf8(Command::new("node").arg("-v").output().unwrap().stdout).unwrap();
+
+    let temp_dir = PathBuf::from("C:\\Windows\\Temp\\");
```
Comment (sehz) [inline nj-build/src/lib.rs:17]: can you incorporate these changes I made for another branch to get rid of hard coded temp directory?

```
let node_full_version = String::from_utf8(Command::new("node").arg("-v").output().unwrap().stdout).unwrap();
        
    let temp_lib =   temp_dir().join(format!("node-{}.lib", node_full_version.trim_end()));

    if !temp_lib.exists() {

        let lib_file_download_url = format!(
            "https://nodejs.org/dist/{}/win-x64/node.lib",
            node_full_version
        );

        println!("downloading nodejs: {} to: {:#?}",lib_file_download_url,temp_lib);

        let mut resp =
            reqwest::blocking::get(&lib_file_download_url).expect("Download node.lib file failed");
        let mut node_lib_file = File::create(&temp_lib).unwrap();
        copy(&mut resp, &mut node_lib_file).expect("Save node.lib file failed");
    }
```

**Comment by sehz** [comment]
Also, can you create another issue to create step in the CI to test modules for electron?

File: nj-build/src/lib.rs (hunk 3 above)
Comment (Ryanmtate) [reply nj-build/src/lib.rs:17]: Yeah

**Comment by Ryanmtate** [comment]
> Also, can you create another issue to create step in the CI to test modules for electron?

I've opened https://github.com/infinyon/node-bindgen/issues/64

File: nj-build/src/lib.rs (hunk 3 above)
Comment (Ryanmtate) [reply nj-build/src/lib.rs:17]: do you have a definition for `temp_dir()` somewhere?

Comment (sehz) [reply nj-build/src/lib.rs:17]: use std::env::temp_dir;

**Comment by usagi** [comment]
@Ryanmtate Thanks to your great work. But, I could not build an electron native module using this changes. Please help me if you can.

### Problem

The error happen to me in linking time, but it is not same to #17 .

Run `yarn electron-build-env nj-cli build` in PowerShell, then I got the error:

```
  = note: win_delay_load_hook.lib(win_delay_load_hook.o) : error LNK2005: __pfnDliNotifyHook2 already defined in win_delay_load_hook.o
             Creating library C:\Users\usagi\tmp\ex1\target\debug\deps\ex1_marshalling.dll.lib and object C:\Users\usagi\tmp\ex1\target\debug\deps\ex1_marshalling.dll.exp
          C:\Users\usagi\tmp\ex1\target\debug\deps\ex1_marshalling.dll : fatal error LNK1169: one or more multiply defined symbols found
```

Cargo.toml:

```toml
[build-dependencies]
node-bindgen = { git = "https://github.com/infinyon/node-bindgen", branch="master", features = ["build"] }
[dependencies]
node-bindgen = { git = "https://github.com/infinyon/node-bindgen", branch="master" }
```

Runtime environment check:

Run `yarn electron-build-env echo %npm_config_runtime%` in PowerShell then I got:

```
electron
```

Please tell me my mistakens if you can find. Or, please share your minimal project that work fine.

**Comment by Ryanmtate** [comment]
@usagi Also noticing this error in our CI test. Will review and check what I missed and report back.

**Comment by Ryanmtate** [comment]
@usagi https://github.com/infinyon/node-bindgen/pull/67 Has an incoming fix that I think will resolve this problem.

**Pull request merged**